# Q1
Write a python program which searches all the product under a particular product vertical
from www.amazon.in. The product verticals to be searched will be taken as input from user.
For e.g. If user input is ‘guitar’. Then search for guitars.

In [1]:
import selenium
import pandas as pd
import time
from selenium import webdriver
import urllib.request
import os
from selenium.common.exceptions import ElementClickInterceptedException
from selenium.webdriver.common.keys import Keys

In [2]:
# loading the chrome driver
driver1 = webdriver.Chrome(r"D:\Downloads\IDM Downloads\chromedriver.exe")

# opening webpage
url1 = 'http://www.amazon.in/'
driver1.get(url1)
driver1.maximize_window()

In [3]:
# asking user to search for a product
user_input = input("Enter here to search: ")
search_bar1 = driver1.find_element_by_id("twotabsearchtextbox")
search_bar1.send_keys(user_input)
time.sleep(3)

# pressing search button
search_button = driver1.find_element_by_id("nav-search-submit-button")
search_button.click()

Enter here to search: guitar


For example, I searched for guitar as the product. Any product can be searched

# Q2
In the above question, now scrape the following details of each product listed in first 3 pages
of your search results and save it in a dataframe and csv. In case if any product vertical has
less than 3 pages in search results then scrape all the products available under that product
vertical. Details to be scraped are: "Brand Name", "Name of the Product", "Rating", "No. of
Ratings", "Price", "Return/Exchange", "Expected Delivery", "Availability", "Other Details"
and “Product URL”. In case, if any of the details are missing for any of the product then
replace it by “-“.

In [4]:
from selenium.common.exceptions import NoSuchElementException

In [5]:
# saving the 1st page url
page_url2a = 'https://www.amazon.in/s?k=guitar&ref=nb_sb_noss_1' 

# saving the products links in a list to open one by one for scrapping
url2a = []
product_url2a = driver1.find_elements_by_xpath("//a[@class = 'a-link-normal a-text-normal']")
for i in product_url2a:
    url2a.append(i.get_attribute("href"))

# making list for append all the details from 1st webpage of search results
product2a = []
rating2a = []
n_rating2a = []
price2a = []
delivery2a = []
return2a = []


for i in url2a:
    driver1.get(i)
    time.sleep(3)
    
# scraping product name from 1st page
    product_name2a = driver1.find_element_by_xpath("//span[@id = 'productTitle']")
    product2a.append(product_name2a.text)
    
# scraping total number of ratings from 1st webpage
    try:
        number_ratings2a = driver1.find_element_by_xpath("//span[@id = 'acrCustomerReviewText']")
        n_rating2a.append(number_ratings2a.text.replace(' ratings', ''))
    except NoSuchElementException: # if not present then append '-' character
        n_rating2a.append('-')
        
# scraping price of product from 1st page
    try:
        product_price2a = driver1.find_element_by_xpath("//span[@id = 'priceblock_ourprice']")
        price2a.append(product_price2a.text.replace('₹ ', ""))
    except NoSuchElementException:
        price2a.append('-')

# scraping average rating of product from 1st page    
    try:
        product_rating2a = driver1.find_element_by_xpath("//span[@data-hook = 'rating-out-of-text']")
        rating2a.append(product_rating2a.text.replace(' out of 5', ''))
    except NoSuchElementException:
        rating2a.append('-')

# scraping delivery details from 1st page
    try:
        product_delivery2a = driver1.find_element_by_xpath("//div[@id = 'ddmDeliveryMessage']")
        delivery2a.append(product_delivery2a.text.replace(' Details', ''))
    except NoSuchElementException:
        delivery2a.append('-')
    
# scraping return policy from 1st page
    try:
        product_return2a = driver1.find_element_by_xpath("//div[@id = 'RETURNS_POLICY']")
        return2a.append(product_return2a.text)
    except NoSuchElementException:
        return2a.append('-')
        
time.sleep(1)
driver1.get(page_url2a)

In [6]:
# Similarly scrapping all the information from 2nd page
page_url2b = 'https://www.amazon.in/s?k=guitar&page=2&qid=1623070699&ref=sr_pg_1'

url2b = []
product_url2b = driver1.find_elements_by_xpath("//a[@class = 'a-link-normal a-text-normal']")
for i in product_url2b:
    url2b.append(i.get_attribute("href"))

product2b = []
rating2b = []
n_rating2b = []
price2b = []
delivery2b = []
return2b = []

for i in url2b:
    driver1.get(i)
    time.sleep(3)
    
    product_name2b = driver1.find_element_by_xpath("//span[@id = 'productTitle']")
    product2b.append(product_name2b.text)
    
    try:
        number_ratings2b = driver1.find_element_by_xpath("//span[@id = 'acrCustomerReviewText']")
        n_rating2b.append(number_ratings2b.text.replace(' ratings', ''))
    except NoSuchElementException:
        n_rating2b.append('-')
        
    try:
        product_price2b = driver1.find_element_by_xpath("//span[@id = 'priceblock_ourprice']")
        price2b.append(product_price2b.text.replace('₹ ', ""))
    except NoSuchElementException:
        price2b.append('-')
    
    try:
        product_rating2b = driver1.find_element_by_xpath("//span[@data-hook = 'rating-out-of-text']")
        rating2b.append(product_rating2b.text.replace(' out of 5', ''))
    except NoSuchElementException:
        rating2b.append('-')
    
    try:
        product_delivery2b = driver1.find_element_by_xpath("//div[@id = 'ddmDeliveryMessage']")
        delivery2b.append(product_delivery2b.text.replace(' Details', ''))
    except NoSuchElementException:
        delivery2b.append('-')
    
    try:
        product_return2b = driver1.find_element_by_xpath("//div[@id = 'RETURNS_POLICY']")
        return2b.append(product_return2b.text)
    except NoSuchElementException:
        return2b.append('-')
        
time.sleep(1)
driver1.get(page_url2b)

In [7]:
# scraping all the information from 3rd page
url2c = []
product_url2c = driver1.find_elements_by_xpath("//a[@class = 'a-link-normal a-text-normal']")
for i in product_url2c:
    url2c.append(i.get_attribute("href"))

product2c = []
rating2c = []
n_rating2c = []
price2c = []
delivery2c = []
return2c = []

for i in url2c:
    driver1.get(i)
    time.sleep(3)
    
    product_name2c = driver1.find_element_by_xpath("//span[@id = 'productTitle']")
    product2c.append(product_name2c.text)
    
    try:
        number_ratings2c = driver1.find_element_by_xpath("//span[@id = 'acrCustomerReviewText']")
        n_rating2c.append(number_ratings2c.text.replace(' ratings', ''))
    except NoSuchElementException:
        n_rating2c.append('-')
        
    try:
        product_price2c = driver1.find_element_by_xpath("//span[@id = 'priceblock_ourprice']")
        price2c.append(product_price2c.text.replace('₹ ', ""))
    except NoSuchElementException:
        price2c.append('-')
    
    try:
        product_rating2c = driver1.find_element_by_xpath("//span[@data-hook = 'rating-out-of-text']")
        rating2b.append(product_rating2c.text.replace(' out of 5', ''))
    except NoSuchElementException:
        rating2c.append('-')
    
    try:
        product_delivery2c = driver1.find_element_by_xpath("//div[@id = 'ddmDeliveryMessage']")
        delivery2c.append(product_delivery2c.text.replace(' Details', ''))
    except NoSuchElementException:
        delivery2c.append('-')
    
    try:
        product_return2c = driver1.find_element_by_xpath("//div[@id = 'RETURNS_POLICY']")
        return2c.append(product_return2c.text)
    except NoSuchElementException:
        return2c.append('-')
        
driver1.close()

In [8]:
name2 = []
price2 = []
rating2 = []
nrating2 = []
delivery2 = []
return2 = []

# saving all the information in new lists
name2 = product2a + product2b + product2c
price2 = price2a + price2b + price2c
rating2 = rating2a + rating2b + rating2c 
nrating2 = n_rating2a + n_rating2b + n_rating2c
delivery2 = delivery2a + delivery2b + delivery2c
return2 = return2a + return2b + return2c

print(len(name2), len(price2), len(rating2) ,len(nrating2), len(delivery2), len(return2))

155 155 155 155 155 155


In [9]:
# Making a dataframe of the scraped data
user_searched_product2 = pd.DataFrame() 
user_searched_product2['Product Name'] = name2
user_searched_product2['Price of Product'] = price2
user_searched_product2['Product Rating'] = rating2
user_searched_product2['Number of Ratings'] = nrating2
user_searched_product2['Delivery Details'] = delivery2
user_searched_product2['Return Policy'] = return2
user_searched_product2

Product Name Price of Product  \
0    Juârez Acoustic Guitar, 38 Inch Cutaway, 038C ...         2,499.00   
1    Juarez JRZ38C Acoustic Guitar, 38 Inch Cutaway...         2,499.00   
2    JUAREZ JRZ38C Right Handed Acoustic Guitar (Su...         2,499.00   
3    Kadence Acoustica Series Semi Acoustic Ash Woo...         7,999.00   
4    Juârez Acoustic Guitar, 38 Inch Cutaway, 038C ...         2,499.00   
..                                                 ...              ...   
150  HYMET 4-String Acoustic Guitar Learning Kids T...           789.00   
151  JUAREZ Arpeggio 39 Inch Acoustic Guitar Kit, S...                -   
152  Divinext Audio Interface AMP Guitar Effects Pe...           749.00   
153  JUAREZ Arpéggio 41 Inch Acoustic Guitar Kit, S...                -   
154  JATU STORE String Acoustic Guitar | Adjustable...           729.00   

    Product Rating Number of Ratings                   Delivery Details  \
0              3.9             9,827     FREE delivery: Monday, June 14   
1                4             3,368     FREE delivery: Monday, June 14   
2                4             3,500     FREE delivery: Monday, June 14   
3              3.9               333     FREE delivery: Monday, June 14   
4              3.9             9,827     FREE delivery: Monday, June 14   
..             ...               ...                                ...   
150              -          1 rating  FREE delivery: Wednesday, June 16   
151              -                 3     FREE delivery: Monday, June 14   
152              -               107  FREE delivery: Wednesday, June 16   
153              -                 7    FREE delivery: Tuesday, June 15   
154              -                 -  FREE delivery: Wednesday, June 16   

               Return Policy  
0         7 Days Replacement  
1         7 Days Replacement  
2         7 Days Replacement  
3         7 Days Replacement  
4         7 Days Replacement  
..                       ...  
150        7 Days Returnable  
151       7 Days Replacement  
152  7 Days Replacement Only  
153       7 Days Replacement  
154       7 Days Replacement  

[155 rows x 6 columns]

In [10]:
# Making a csv file of scraped data
user_searched_product2.to_csv("Searched product by user.csv")

# Q3
Write a python program to access the search bar and search button on images.google.com and
scrape 100 images each for keywords ‘fruits’, ‘cars’ and ‘Machine Learning’.

In [2]:
driver3 = webdriver.Chrome(r"D:\Downloads\IDM Downloads\chromedriver.exe")
url3 = "https://www.google.com/imghp"
driver3.get(url3)
driver3.maximize_window()

In [3]:
# asking the user for keyword to search and number of images.
search_input3 = input("What images do you want to search: ")
search_num3 = int(input("How many images do you want: "))

search_bar = driver3.find_element_by_xpath("//input[@class = 'gLFyf gsfi']")
search_bar.send_keys(search_input3)

search_button = driver3.find_element_by_xpath("//span[@class = 'z1asCe MZy1Rb']")
search_button.click()

What images do you want to search: fruits
How many images do you want: 100


Now we scroll to the end of the page to avoid the exceptions

In [4]:
scroll_pause_time = 2

# Get scroll height
last_height = driver3.execute_script("return document.body.scrollHeight")

while True:
    # Scroll down to bottom
    driver3.execute_script("window.scrollTo(0, document.body.scrollHeight);")

    # Wait to load page
    time.sleep(scroll_pause_time)

    # Calculate new scroll height and compare with last scroll height
    new_height = driver3.execute_script("return document.body.scrollHeight")
    if new_height == last_height:
        break
    last_height = new_height
    
# to scroll to the top of the page.
from selenium.webdriver.common.keys import Keys
driver3.find_element_by_tag_name('body').send_keys(Keys.CONTROL + Keys.HOME)

In [5]:
url3a = []
# searching for the links of all the fruits images in webelements
fruit_img = driver3.find_elements_by_xpath("//img[@class = 'rg_i Q4LuWd']")
for i in range(search_num3):
# appending the links in the list    
    url3a.append(fruit_img[i].get_attribute("src"))

driver3.get(url3)
time.sleep(3)

search_input3 = input("What images do you want to search: ")
search_num3 = int(input("How many images do you want: "))

search_bar = driver3.find_element_by_xpath("//input[@class = 'gLFyf gsfi']")
search_bar.send_keys(search_input3)

search_button = driver3.find_element_by_xpath("//span[@class = 'z1asCe MZy1Rb']")
search_button.click()

What images do you want to search: cars
How many images do you want: 100


In [6]:
scroll_pause_time = 2

# Get scroll height
last_height = driver3.execute_script("return document.body.scrollHeight")

while True:
    # Scroll down to bottom
    driver3.execute_script("window.scrollTo(0, document.body.scrollHeight);")

    # Wait to load page
    time.sleep(scroll_pause_time)

    # Calculate new scroll height and compare with last scroll height
    new_height = driver3.execute_script("return document.body.scrollHeight")
    if new_height == last_height:
        break
    last_height = new_height
    
# to scroll to the top of the page.
from selenium.webdriver.common.keys import Keys
driver3.find_element_by_tag_name('body').send_keys(Keys.CONTROL + Keys.HOME)

In [7]:
url3b = []
# searching for the links of all the cars images in webelements
cars_img = driver3.find_elements_by_xpath("//img[@class = 'rg_i Q4LuWd']")
for i in range(search_num3):
# appending the links in the list    
    url3b.append(cars_img[i].get_attribute("src"))

driver3.get(url3)
time.sleep(3)

search_input3 = input("What images do you want to search: ")
search_num3 = int(input("How many images do you want: "))

search_bar = driver3.find_element_by_xpath("//input[@class = 'gLFyf gsfi']")
search_bar.send_keys(search_input3)

search_button = driver3.find_element_by_xpath("//span[@class = 'z1asCe MZy1Rb']")
search_button.click()

What images do you want to search: machine learning
How many images do you want: 100


In [8]:
scroll_pause_time = 2

# Get scroll height
last_height = driver3.execute_script("return document.body.scrollHeight")

while True:
    # Scroll down to bottom
    driver3.execute_script("window.scrollTo(0, document.body.scrollHeight);")

    # Wait to load page
    time.sleep(scroll_pause_time)

    # Calculate new scroll height and compare with last scroll height
    new_height = driver3.execute_script("return document.body.scrollHeight")
    if new_height == last_height:
        break
    last_height = new_height
    
# to scroll to the top of the page.
from selenium.webdriver.common.keys import Keys
driver3.find_element_by_tag_name('body').send_keys(Keys.CONTROL + Keys.HOME)

In [9]:
url3c = []
# searching for the links of all the machine learning images in webelements
ml_img = driver3.find_elements_by_xpath("//img[@class = 'rg_i Q4LuWd']")
for i in range(search_num3):
# appending the links in the list    
    url3c.append(ml_img[i].get_attribute("src"))

driver3.close()

In [13]:
# making a function to doenload the images.
def img_download(urls : list, search_input : str):
    new_dir = "Downloaded_images" '''to make a new folder named Downloaded_images'''
    parent_dir = "D:\\" '''Downloaded_images folder to be made into D drive'''
    path = os.path.join(parent_dir, new_dir) '''joining both folder and D drive using path-join'''
    if not os.path.exists("D:\Downloaded_images"): '''create folder only if it does not exist'''
        os.mkdir(path) '''make the folder'''
    os.chdir(os.path.join(os.getcwd(), "D:\Downloaded_images")) '''change current directory to given path using path-join'''
    for i in range(search_num3): '''taking number of images wanted by user as range'''
        try: '''try to download the image in the url from 'urls' list one by one'''  
            urllib.request.urlretrieve(urls[i], f"{search_input} {i+1}.jpg") 
            print(f"image saved as {search_input} {i+1}.jpg")
        except Exception as e: '''if an error occurs, except it then print the error.'''
            print(f"ERROR - {e}")

In [14]:
img_download(url3a, 'fruits')

image saved as fruits 1.jpg
image saved as fruits 2.jpg
image saved as fruits 3.jpg
image saved as fruits 4.jpg
image saved as fruits 5.jpg
image saved as fruits 6.jpg
image saved as fruits 7.jpg
image saved as fruits 8.jpg
image saved as fruits 9.jpg
image saved as fruits 10.jpg
image saved as fruits 11.jpg
image saved as fruits 12.jpg
image saved as fruits 13.jpg
image saved as fruits 14.jpg
image saved as fruits 15.jpg
image saved as fruits 16.jpg
image saved as fruits 17.jpg
image saved as fruits 18.jpg
image saved as fruits 19.jpg
image saved as fruits 20.jpg
image saved as fruits 21.jpg
image saved as fruits 22.jpg
image saved as fruits 23.jpg
image saved as fruits 24.jpg
image saved as fruits 25.jpg
image saved as fruits 26.jpg
image saved as fruits 27.jpg
image saved as fruits 28.jpg
image saved as fruits 29.jpg
image saved as fruits 30.jpg
image saved as fruits 31.jpg
image saved as fruits 32.jpg
image saved as fruits 33.jpg
image saved as fruits 34.jpg
image saved as fruits 3

In [15]:
img_download(url3b, 'cars')

image saved as cars 1.jpg
image saved as cars 2.jpg
image saved as cars 3.jpg
image saved as cars 4.jpg
image saved as cars 5.jpg
image saved as cars 6.jpg
image saved as cars 7.jpg
image saved as cars 8.jpg
image saved as cars 9.jpg
image saved as cars 10.jpg
image saved as cars 11.jpg
image saved as cars 12.jpg
image saved as cars 13.jpg
image saved as cars 14.jpg
image saved as cars 15.jpg
image saved as cars 16.jpg
image saved as cars 17.jpg
image saved as cars 18.jpg
image saved as cars 19.jpg
image saved as cars 20.jpg
image saved as cars 21.jpg
image saved as cars 22.jpg
image saved as cars 23.jpg
image saved as cars 24.jpg
image saved as cars 25.jpg
image saved as cars 26.jpg
image saved as cars 27.jpg
image saved as cars 28.jpg
image saved as cars 29.jpg
image saved as cars 30.jpg
image saved as cars 31.jpg
image saved as cars 32.jpg
image saved as cars 33.jpg
image saved as cars 34.jpg
image saved as cars 35.jpg
image saved as cars 36.jpg
image saved as cars 37.jpg
image save

In [16]:
img_download(url3c, 'machine learning')

image saved as machine learning 1.jpg
image saved as machine learning 2.jpg
image saved as machine learning 3.jpg
image saved as machine learning 4.jpg
image saved as machine learning 5.jpg
image saved as machine learning 6.jpg
image saved as machine learning 7.jpg
image saved as machine learning 8.jpg
image saved as machine learning 9.jpg
image saved as machine learning 10.jpg
image saved as machine learning 11.jpg
image saved as machine learning 12.jpg
image saved as machine learning 13.jpg
image saved as machine learning 14.jpg
image saved as machine learning 15.jpg
image saved as machine learning 16.jpg
image saved as machine learning 17.jpg
image saved as machine learning 18.jpg
image saved as machine learning 19.jpg
image saved as machine learning 20.jpg
image saved as machine learning 21.jpg
image saved as machine learning 22.jpg
image saved as machine learning 23.jpg
image saved as machine learning 24.jpg
image saved as machine learning 25.jpg
image saved as machine learning 26

# Q4
Write a python program to search for a smartphone(e.g.: Oneplus Nord, pixel 4A, etc.) on
www.flipkart.com and scrape following details for all the search results displayed on 1 st page.
Details to be scraped: “Brand Name”, “Smartphone name”, “Colour”, “RAM”,
“Storage(ROM)”, “Primary Camera”, “Secondary Camera”, “Display Size”, “Display
Resolution”, “Processor”, “Processor Cores”, “Battery Capacity”, “Price”, “Product URL”.
Incase if any of the details is missing then replace it by “- “. Save your results in a dataframe
and CSV.

In [17]:
driver_path = r"D:\Downloads\IDM Downloads\chromedriver.exe"
driver4 = webdriver.Chrome(driver_path) # opening the website
url4 = 'http://www.flipkart.com/'
driver4.get(url4)
driver4.maximize_window()

cross_button = driver4.find_element_by_xpath("//button[@class='_2KpZ6l _2doB4z']")
cross_button.click()

In [18]:
# any smartphone can be searched, so searching for 'Iqoo 3'
search_bar = driver4.find_element_by_xpath("//input[@class = '_3704LK']")
search_bar.click()
search_bar.send_keys('iqoo 3')
time.sleep(0.5)

search_button = driver4.find_element_by_xpath("//button[@class = 'L0Z3Pu']")
search_button.click()

In [19]:
# first finding the urls for each smartphone given in the webpage.
products4 = driver4.find_elements_by_xpath("//a[@class='_1fQZEK']")
urls4 = []
for i in products4:
    urls4.append(i.get_attribute("href"))
len(urls4)

5

As we see we have a total of 5 phone on the webpage, so we scrap all the details for each of these 5 smartphones.

In [20]:
name4 = [] # making empty lists to append name of smartphone
price4 = [] # making empty lists to append price of smartphone
elements = [] # making empty lists to append other details of smartphone

for i in urls4: # loop to open each url in list 'urls4' one by one, then scrap all the details
    driver4.get(i)
    time.sleep(2)
    read_more_button = driver4.find_element_by_xpath("//button[@class = '_2KpZ6l _1FH0tX']")
    read_more_button.click()
    time.sleep(0.5)
    
    phone_name4 = driver4.find_element_by_xpath("//span[@class = 'B_NuCI']")
    phone_price4 = driver4.find_element_by_xpath("//div[@class = '_30jeq3 _16Jk6d']")
    element = driver4.find_elements_by_xpath("//li[@class = '_21lJbe']")
    
    
    name4.append(phone_name4.text) # appending names in the list
    price4.append(phone_price4.text) # appending price in the list
    for j in element:
        elements.append(j.text) # appending all the other elements in the list

driver4.close()

print(len(name4), len(price4), len(elements))

5 5 370


5 divided by the 370 which is length of elements list gives value 74. So first 74 values are the details of 1st smartphone, similarly, second 74 values are the details of second smartphone and so on. We will use the list slicing to get the each detail seperately.

In [22]:
for i in range(len(elements)):
    print(f"{i} - {elements[i]}")

0 - Handset, Earphone, Type-C USB to USB Cable, USB Power Adapter, SIM Ejector, Protective Case, Protective Film (Applied), Documentation
1 - I1927
2 - 3
3 - Quantum Silver
4 - Smartphones
5 - Dual Sim
6 - No
7 - Yes
8 - Yes
9 - Yes
10 - Head: 1.190 W/kg, Body: 0.590 W/kg
11 - 16.36 cm (6.44 inch)
12 - 2400 x 1080 Pixels
13 - Full HD+
14 - Adreno 650
15 - Super AMOLED
16 - 16M
17 - Screen HDR: HDR 10+, Bezel: 5.17mm (Lower), Screen to Body Ratio: 91.40%, Backside Corning Gorilla Glass, Corning Gorilla Glass Gen: GG6, Monster Touch Buttons 180Hz Touch Response Rate, Material Display: Schott Xensation UP
18 - Android 10
19 - Qualcomm Snapdragon 865
20 - Octa Core
21 - 2.8 GHz
22 - 2.4 GHz
23 - 1.8 GHz
24 - 2G GSM: B2/B3/B5/B8, 3G WCDMA: B1/B2/B4/B5/B8, 4G TDD_LTE: B34/B38/B39/B40/B41, 4G FDD_LTE: B1/B2/B3/B4/B5/B7/B8/B12/B17/B18/B19/B20/B25/B26, 4G+: B1/B3/B5/B8/B39/B40/B41/38
25 - 128 GB
26 - 8 GB
27 - Yes
28 - 48MP + 13MP + 13MP + 2MP
29 - 48MP + 13MP (Telephoto - 20X Digital Zoom) + 1

from the above function, we see that values along with the index for each details. So we use the above index for the slicing of list.

In [23]:
# slicing the lists according to respective indexes to get the details.
colour4 = elements[3::74]
RAM4 = elements[26::74]
storage4 = elements[25::74]
primary_camera4 = elements[28::74]
secondary_camera4 = elements[31::74]
display_size4 = elements[11::74]
display_resolution4 = elements[12::74]
processor4 = elements[19::74]
processor_core4 = elements[20::74]
battery4 = elements[69::74]

In [24]:
print(name4)
print(price4)
print(colour4)
print(RAM4)
print(storage4)
print(primary_camera4)
print(secondary_camera4)
print(display_resolution4)
print(processor4)
print(processor_core4) 
print(battery4)

['IQOO 3 (Quantum Silver, 128 GB)  (8 GB RAM)', 'IQOO 3 (Quantum Silver, 256 GB)  (8 GB RAM)', 'IQOO 3 (Tornado Black, 256 GB)  (8 GB RAM)', 'IQOO 3 (Tornado Black, 128 GB)  (8 GB RAM)', 'IQOO 3 (5G) (Tornado Black, 256 GB)  (12 GB RAM)']
['₹34,990', '₹37,990', '₹37,990', '₹34,990', '₹39,990']
['Quantum Silver', 'Quantum Silver', 'Tornado Black', 'Tornado Black', 'Tornado Black']
['8 GB', '8 GB', '8 GB', '8 GB', '12 GB']
['128 GB', '256 GB', '256 GB', '128 GB', '256 GB']
['48MP + 13MP + 13MP + 2MP', '48MP + 13MP + 13MP + 2MP', '48MP + 13MP + 13MP + 2MP', '48MP + 13MP + 13MP + 2MP', '48MP + 13MP + 13MP + 2MP']
['16MP Front Camera', '16MP Front Camera', '16MP Front Camera', '16MP Front Camera', '16MP Front Camera']
['2400 x 1080 Pixels', '2400 x 1080 Pixels', '2400 x 1080 Pixels', '2400 x 1080 Pixels', '2400 x 1080 Pixels']
['Qualcomm Snapdragon 865', 'Qualcomm Snapdragon 865', 'Qualcomm Snapdragon 865', 'Qualcomm Snapdragon 865', 'Qualcomm Snapdragon 865']
['Octa Core', 'Octa Core', 'Oc

As we see that we have scrapped all the necessary details. We will now make a dataframe and save the values in csv.

In [26]:
searched_smartphone = pd.DataFrame()
searched_smartphone['Phone_name'] = name4
searched_smartphone['Phone_colour'] = colour4
searched_smartphone['RAM'] = RAM4
searched_smartphone['Storage'] = storage4
searched_smartphone['Primary_camera'] = primary_camera4
searched_smartphone['Secondary_camera'] = secondary_camera4
searched_smartphone['Display_size'] = display_size4
searched_smartphone['Display_resolution'] = display_resolution4
searched_smartphone['Processor'] = processor4
searched_smartphone['Processor_core'] = processor_core4
searched_smartphone['Battery_capacity'] = battery4
searched_smartphone['Price'] = price4
searched_smartphone['Phone_URL'] = urls4
searched_smartphone['Brand_name'] = 'IQOO 3'
searched_smartphone

Phone_name    Phone_colour    RAM  \
0       IQOO 3 (Quantum Silver, 128 GB)  (8 GB RAM)  Quantum Silver   8 GB   
1       IQOO 3 (Quantum Silver, 256 GB)  (8 GB RAM)  Quantum Silver   8 GB   
2        IQOO 3 (Tornado Black, 256 GB)  (8 GB RAM)   Tornado Black   8 GB   
3        IQOO 3 (Tornado Black, 128 GB)  (8 GB RAM)   Tornado Black   8 GB   
4  IQOO 3 (5G) (Tornado Black, 256 GB)  (12 GB RAM)   Tornado Black  12 GB   

  Storage            Primary_camera   Secondary_camera          Display_size  \
0  128 GB  48MP + 13MP + 13MP + 2MP  16MP Front Camera  16.36 cm (6.44 inch)   
1  256 GB  48MP + 13MP + 13MP + 2MP  16MP Front Camera  16.36 cm (6.44 inch)   
2  256 GB  48MP + 13MP + 13MP + 2MP  16MP Front Camera  16.36 cm (6.44 inch)   
3  128 GB  48MP + 13MP + 13MP + 2MP  16MP Front Camera  16.36 cm (6.44 inch)   
4  256 GB  48MP + 13MP + 13MP + 2MP  16MP Front Camera  16.36 cm (6.44 inch)   

   Display_resolution                Processor Processor_core  \
0  2400 x 1080 Pixels  Qualcomm Snapdragon 865      Octa Core   
1  2400 x 1080 Pixels  Qualcomm Snapdragon 865      Octa Core   
2  2400 x 1080 Pixels  Qualcomm Snapdragon 865      Octa Core   
3  2400 x 1080 Pixels  Qualcomm Snapdragon 865      Octa Core   
4  2400 x 1080 Pixels  Qualcomm Snapdragon 865      Octa Core   

  Battery_capacity    Price  \
0         4440 mAh  ₹34,990   
1         4440 mAh  ₹37,990   
2         4440 mAh  ₹37,990   
3         4440 mAh  ₹34,990   
4         4440 mAh  ₹39,990   

                                           Phone_URL Brand_name  
0  https://www.flipkart.com/iqoo-3-quantum-silver...     IQOO 3  
1  https://www.flipkart.com/iqoo-3-quantum-silver...     IQOO 3  
2  https://www.flipkart.com/iqoo-3-tornado-black-...     IQOO 3  
3  https://www.flipkart.com/iqoo-3-tornado-black-...     IQOO 3  
4  https://www.flipkart.com/iqoo-3-5g-tornado-bla...     IQOO 3

In [27]:
searched_smartphone.to_csv("Samrtphone_results_flipkart.csv")

# Q5
Write a program to scrap geospatial coordinates (latitude, longitude) of a city searched on
google maps. For this problem, we will ask the user to search for the city and then give the coordinates.

In [13]:
# Defining a function to find the coordinates for any city.
def find_coordinates(city : str):

    """Takes city as the input in string format that we want to search"""
    
    driver5 = webdriver.Chrome(driver_path)
    url5 = 'https://www.google.com/maps'
    driver5.get(url5)
    driver5.maximize_window()
    time.sleep(0.5)
    
    search_bar5 = driver5.find_element_by_xpath("//input[@aria-label = 'Search Google Maps']")
    search_bar5.send_keys(city, Keys.ENTER)
    time.sleep(2.5)

    city_search_url = driver5.current_url
    driver5.close()
    
    print(f"The map url for searched city is\n{city_search_url}\n")

    coordinates = city_search_url.split('/')
    coordinates = coordinates[6]
    coordinates = coordinates.replace('@', '').replace('z', '')
    coordinates = coordinates.split(',')
    latitude = coordinates[0]
    longitude = coordinates[1]
    print(f"Longitute of {city} is: {longitude}\nLatitude of {city} is: {latitude}")

Now we will use the above function to scrap the coordinates for any city in the world. It will take the city that we want to search as the input.

In [11]:
find_coordinates('Durg')

The map url for searched city is
https://www.google.com/maps/search/Durg/@21.1923751,81.2479011,13z

Longitute of Durg is: 81.2479011
Latitude of Durg is: 21.1923751


In [12]:
find_coordinates('Nagpur')

The map url for searched city is
https://www.google.com/maps/place/Nagpur,+Maharashtra/@21.1610714,79.0024695,12z/data=!3m1!4b1!4m5!3m4!1s0x3bd4c0a5a31faf13:0x19b37d06d0bb3e2b!8m2!3d21.1458004!4d79.0881546

Longitute of Nagpur is: 79.0024695
Latitude of Nagpur is: 21.1610714


# Q6
Write a program to scrap details of all the funding deals for second quarter (i.e. July 20 –
September 20) from trak.in.

In [21]:
driver6 = webdriver.Chrome(driver_path)
url6 = 'https://trak.in/'
driver6.get(url6)
driver6.maximize_window()
time.sleep(1)

open_funding_deals6 = driver6.find_element_by_id("menu-item-51510")
open_funding_deals6.click()

After inspecting the tables for July 2020 to September 2020, we see that each have unique id, so we use it to scrap the data.

In [22]:
# Making empty lists to append the scraped data.
date6a = []
startupname6a = []
industryvertical6a = []
subvertical6a = []
location6a = []
investorname6a = []
investmenttype6a = []
amount6a = []


for i in range(48,51):
    # To scrap the date Date
    dates6 = driver6.find_elements_by_xpath('//table[@id="tablepress-{}"]/tbody/tr/td[2]'.format(i))
    for j in dates6:
        date6a.append(j.text)

    # To scrap the Startup Name
    startup_names6 = driver6.find_elements_by_xpath('//table[@id="tablepress-{}"]/tbody/tr/td[3]'.format(i))
    for j in startup_names6:
        startupname6a.append(j.text)
    
    # To scrap the Industry/Vertical
    industry_vertical6 = driver6.find_elements_by_xpath('//table[@id="tablepress-{}"]/tbody/tr/td[4]'.format(i))
    for j in industry_vertical6:
        industryvertical6a.append(j.text)
    
    # To scrap the Sub-Vertical
    sub_vertical6 = driver6.find_elements_by_xpath('//table[@id="tablepress-{}"]/tbody/tr/td[5]'.format(i))
    for j in sub_vertical6:
        subvertical6a.append(j.text)

    # To scrap the Location
    location6 = driver6.find_elements_by_xpath('//table[@id="tablepress-{}"]/tbody/tr/td[6]'.format(i))
    for j in location6:
        location6a.append(j.text)
    
    # To scrap the Investor
    investor_name6 = driver6.find_elements_by_xpath('//table[@id="tablepress-{}"]/tbody/tr/td[7]'.format(i))
    for j in investor_name6:
        investorname6a.append(j.text)
    
    # To scrap the Investment Type
    investment_type6 = driver6.find_elements_by_xpath('//table[@id="tablepress-{}"]/tbody/tr/td[8]'.format(i))
    for j in investment_type6:
        investmenttype6a.append(j.text)
    
    # To scrap the Amount
    amount6 = driver6.find_elements_by_xpath('//table[@id="tablepress-{}"]/tbody/tr/td[9]'.format(i))
    for j in amount6:
        amount6a.append(j.text)
        
driver6.close()

In [23]:
funding_deals = pd.DataFrame()
funding_deals['Date'] = date6a
funding_deals['Startup Name'] = startupname6a
funding_deals['Industry Vertical'] = industryvertical6a
funding_deals['Sub-Vertical'] = subvertical6a
funding_deals['City/Location'] = location6a
funding_deals["Investor's Name"] = investorname6a
funding_deals['Investment Type'] = investmenttype6a
funding_deals['Amount(in $)'] = amount6a
funding_deals

Date               Startup Name  \
0   15/07/2020                   Flipkart   
1   16/07/2020                    Vedantu   
2   16/07/2020                       Crio   
3   14/07/2020                    goDutch   
4   13/07/2020                   Mystifly   
5   09/07/2020               JetSynthesys   
6   10/07/2020                   gigIndia   
7   15/07/2020                  PumPumPum   
8   14/07/2020                       FLYX   
9   13/07/2020  Open Appliances Pvt. Ltd.   
10  15/08/2020                     Practo   
11  13/08/2020                    Medlife   
12  13/08/2020                  HungerBox   
13  04/08/2020                      Dunzo   
14  11/08/2020                   Terra.do   
15  12/08/2020                  Classplus   
16  14/08/2020                       Niyo   
17  10/08/2020                  ZestMoney   
18  07/08/2020                FreshToHome   
19  13/08/2020                    Eduvanz   
20  08/09/2020                     Byju’s   
21  12/09/2020                  mCaffeine   
22  09/09/2020                     Qshala   
23  02/09/2020                      Winzo   
24  09/09/2020                Hippo Video   
25  07/09/2020                    Melorra   
26  07/09/2020                        1mg   
27  31/08/2020                      mfine   
28  31/08/2020                       Apna   
29  03/09/2020                    Railofy   

                         Industry Vertical  \
0                               E-commerce   
1                                  EduTech   
2                                  EduTech   
3                                  FinTech   
4                      Airfare Marketplace   
5                 Gaming and Entertainment   
6                              Marketplace   
7                        Automotive Rental   
8                               OTT Player   
9                   Information Technology   
10                              HealthTech   
11                              E-commerce   
12                                FoodTech   
13                   Hyper-local Logistics   
14                                 EduTech   
15                                 EduTech   
16                                 FinTech   
17                                 FinTech   
18                              E-commerce   
19                                 FinTech   
20                                 EduTech   
21                           Personal Care   
22                                 EduTech   
23                           Online Gaming   
24  Video Customer Experience(CX) Platform   
25                              E-commerce   
26                              E-commerce   
27                              HealthTech   
28                         Human Resources   
29                          Transportation   

                                         Sub-Vertical  \
0                                          E-commerce   
1                                     Online Tutoring   
2                    Learning Platform for Developers   
3                                      Group Payments   
4   Ticketing, Airline Retailing, and Post-Ticketi...   
5                            Gaming and Entertainment   
6                           Crowd Sourcing, Freelance   
7                           Used Car-leasing platform   
8                            Streaming Social Network   
9               Internet-of-Things Security Solutions   
10                           Health care and Wellness   
11                                    Online Pharmacy   
12                       Online Food Delivery Service   
13                           Online Delivery Services   
14                  Online Climate School, E-learning   
15                        E-learning, Online Tutoring   
16                                 Financial Services   
17                                 Financial Services   
18                                      Food Delivery   
19                                 Financial Se

# Q7
Write a program to scrap all the available details of top 10 gaming laptops from digit.in.

In [24]:
url7 = 'https://www.digit.in/'
driver7 = webdriver.Chrome(driver_path)
driver7.get(url7)
driver7.maximize_window()
time.sleep(1)

# to go to top 10 page
top_10_button = driver7.find_element_by_xpath("/html/body/div[1]/div[2]/div[4]/ul/li[4]/a")
top_10_button.click()
time.sleep(1)

# to go to laptop page
laptop_selection = driver7.find_element_by_xpath("/html/body/div[3]/div/div/div[2]/div[5]/div[1]/div/button[2]")
laptop_selection.click()
time.sleep(1)

# to go to gaming laptop page
gaming_laptop_selection = driver7.find_element_by_xpath('//*[@id="laptops"]/div[3]/a/div[2]')
gaming_laptop_selection.click()

In [25]:
# Scraping for the names of the laptops
name7 = []
names = driver7.find_elements_by_xpath("//div[@data-eventaction = 'NonAMPScroll']")
for i in names:
    name7.append(i.text.replace('\n', ''))
#name7

# scraping the operating system of laptops
os7 = []
operatingsystem = driver7.find_elements_by_xpath("//div[@class = 'Spcs-details']//tr[3]//td[3]")
for i in operatingsystem:
    os7.append(i.text)
#os7

# scraping the display resolution of laptops
display7 = []
display = driver7.find_elements_by_xpath("//div[@class = 'Spcs-details']//tr[4]//td[3]")
for i in display:
    display7.append(i.text)
#display7

# scraping the processor of laptops 
processor7 = []
processor = driver7.find_elements_by_xpath("//div[@class = 'Spcs-details']//tr[5]//td[3]")
for i in processor:
    processor7.append(i.text)
#processor7

# scraping the memory of laptops
memory7 = []
memory = driver7.find_elements_by_xpath("//div[@class = 'Spcs-details']//tr[6]//td[3]")
for i in memory:
    memory7.append(i.text)
#memory7

# scraping the weight of laptops
weight7 = []
weight = driver7.find_elements_by_xpath("//div[@class = 'Spcs-details']//tr[7]//td[3]")
for i in weight:
    weight7.append(i.text)
#weight7

# scraping the dimension of laptops
dimension7 = []
dimension = driver7.find_elements_by_xpath("//div[@class = 'Spcs-details']//tr[8]//td[3]")
for i in dimension:
    dimension7.append(i.text)
#dimension7

# scraping the price of laptops
price7 = []
price = driver7.find_elements_by_xpath("//td[@class = 'smprice']")
for i in price:
    price7.append(i.text)
#price7

driver7.close()

In [27]:
top_gaming_laptops = pd.DataFrame()
top_gaming_laptops['Name'] = name7
top_gaming_laptops['Operating System'] = os7
top_gaming_laptops['Display'] = display7
top_gaming_laptops['Processor'] = processor7
top_gaming_laptops['Memory Storage'] = memory7
top_gaming_laptops['Weight'] = weight7
top_gaming_laptops['Dimension'] = dimension7
top_gaming_laptops['Price'] = price7
top_gaming_laptops

Name Operating System              Display  \
0     1.ALIENWARE AREA 51M R2  Windows 10 Home  17.3" (1920 x 1080)   
1          2.ALIENWARE M15 R3  Windows 10 Home  15.6" (3840 x 2160)   
2    3.ASUS ROG STRIX SCAR 15  Windows 10 Home  15.6" (1920 x 1080)   
3     4.ASUS ROG ZEPHYRUS G14  Windows 10 Home    14" (1920 x 1080)   
4          5.LENOVO LEGION 5I   Windows 10 Pro  15.6" (1920 x 1080)   
5  6.ASUS ROG ZEPHYRUS DUO 15       Windows 10  15.6" (3840 x 1100)   
6      7.ACER ASPIRE 7 GAMING  Windows 10 Home  15.6" (1920 x 1080)   

                                           Processor          Memory Storage  \
0   10th Generation Intel® Core™ i7-10700 | 2.90 GHz   1 TB SSD/16 GBGB DDR4   
1       10th Generation Intel® Core™ i9-10980HK | NA   1 TB SSD/16 GBGB DDR4   
2                      AMD Ryzen™ 9 5900HX | 3.3 GHz   1 TB SSD/16 GBGB DDR4   
3               AMD 3rd Generation Ryzen 9 | 3.3 GHz   1 TB SSD/16 GBGB DDR4   
4  10th Generation Intel® Core™ i5-10300H | 2.50 GHz   1 TB SSD/16 GBGB DDR4   
5                 Intel Core i7 10th Gen 10875H | NA  512 GB SSD/4 GBGB DDR4   
6                  AMD Ryzen™ 5-5500U hexa-core | NA  512 GB SSD/8 GBGB DDR4   

  Weight                Dimension    Price  
0    4.1   27.65 x 402.6 x 319.14      N/A  
1     NA                       NA  ₹341990  
2   2.30       35.4 x 25.9 x 2.26      N/A  
3   1.65        32.5 x 22.1 x 1.8  ₹164990  
4    2.3  363.06 x 259.61 x 23.57   ₹71990  
5    2.4  268.30 x 360.00 x 20.90  ₹199990  
6   2.15       2.29 x 36.3 x 25.4   ₹56990